### PR TITLE
Bug/282 player join during selection phase

### LINKED ISSUE #282: [BUG]: Player joining during selection phase causes game to get stuck
**Device Used:** N/A

**Browser(s) Used:**
- [x] Chrome
- [x] Safari
- [x] FireFox
- [x] Other: All

**Environment:** 
- [x] Production
- [ ] Staging (N/A)
- [x] Localhost

# Description

### Expected Behavior
A player joining the game during the Player submission phase should get a hand with random cards so they can play the game as expected. 

### Actual Behavior
The game gets "stuck" when a new player joins during the submission phase, not allowing the joined player to send any cards but the game is expecting cards, therefore it gets stuck.

# Screen Shot/Screen Video Capture 
(if applicable)

# Replication Steps & Data
1. Setup a game with 3 players
2. Play the game normally until the Player submission phase
3. Add a new player before submitting all players cards
4. Submit all of the player's cards
5. Observe that the game will not progress

# User Story
As a user, I would like to join in the middle of a game without causing the game to get stuck so when I'm late to the party, I can still be a part of the laughs.

# Acceptance Criteria
Criteria needed for this to work as intended in the user story.
- [x] New player gets a hand of cards and the game waits for them to submit.
- [x] Game does not get stuck on the submission phase.
- [x] Game progresses as normal after player has joined.

### PR DESCRIPTION
**Pull request by fortypercenttitanium**
#### 0. Make sure your issue is linked:

Close #282 

#### 1. Purpose:

Currently if a player joins mid-game while other players are selecting their cards, the game can get stuck waiting for the new player to submit cards they don't have. This fix will deal new cards to the new player and allow them to play right away, thus keeping the game moving.

#### 2. Implementation:

This one required a little bit of a rework in multiple areas:

### HostReducer
* PLAYER_CONNECTED
  * Check if the game state is waiting for players to submit cards
    * If so, deal cards again 
      * current players will still only have the cards they previously had, the new one will receive a full hand
    * If not, proceed as normal
  * Decouples clearing submitted cards from the `dealWhiteCards` function so we can retain which players have submitted cards
  * Imports hostReducerMiddleware to send the cards to the players again once they've been updated. This function is only a side effect and does not fallthrough to a dispatch, so it is a safe import.
    * When sending cards, implements a `newPlayer` property to indicate whether the cards are being dealt as a result of a new player joining. This ensures that the middleware sends the correct socket message to the players (see next).

### hostReducerMiddleware
* `sendCardsToPlayers`
  * When sending cards, implements a `shouldPreserveGameState` property to indicate whether the players receiving cards should be brought to the player-select screen. This boolean (calculated based on the new player's ID compared to the player receiving cards) ensures that players who have already submitted their cards aren't brought back to the selection screen.

### PlayerReducer
Looks to the payload to indicate whether the player's game state should change to 'player-select' (see above)

### HostScreenController
Removes a game state case that should not have been added to the host end. 

#### 3. Possible Issues:

Consider additional edge cases related to the game state and new players joining. Try joining with multiple players mid game.

#### 4. How to Test:

1. Pull branch
2. Run game, open 5+ tabs
3. Join with 3 players, start game, submit cards on one of the tabs
4. Join game with an additional player (or more).
5. Observe the game state of each player (and host), attempt to continue on with the game.

#### 5. Cleanup
- [x] I've added all TODOs needed
- [x] I've removed all unneeded comments
- [x] I've updated the snapshot tests and checked to make sure they're accurate if they changed
- [x] I've removed all unnecessary `console.log`s

### CODE REVIEW

**Review by I3uckwheat — DISMISSED** [review]

**Review by I3uckwheat — CHANGES_REQUESTED** [review]

**Review by pklepa — COMMENTED** [review]
I tested this implementation in different scenarios: one person joining whilst on `waiting-to-receive-cards`, multiple people joining, a person joining whilst another leaves...

Although it seems to be solid and I didn't manage to reproduce the loop that Briggs mentioned, I do agree that moving the conditional (whether or not the game is running) and its subsequent side effects over to the event level in `HostContext` would be beneficial

**Review by I3uckwheat — CHANGES_REQUESTED** [review]

File: frontend/src/components/PlayerList/PlayerList.js [hunk 1]
```diff
@@ -85,7 +100,10 @@ const PlayerRow = styled.div`
   }
 `;
 function PlayerList({ playerList }) {
-  const playersArray = playerList.playerIDs.map((id) => playerList.players[id]);
+  const playersArray = [
+    ...playerList.playerIDs.map((id) => playerList.players[id]),
+    ...playerList.newPlayerStaging,
+  ];
```
Comment (I3uckwheat) [inline frontend/src/components/PlayerList/PlayerList.js:106]: This will show staged and in-play players, right?

File: frontend/src/screenControllers/HostScreenController/HostScreenController.spec.js [hunk 2]
```diff
@@ -99,23 +99,6 @@ describe('Host screen controller', () => {
...
-        render(
-          <HostContext.Provider value={{ state, dispatch }}>
-            <HostScreenController />
-          </HostContext.Provider>,
-        );
-
-        expect(screen.getByTestId('host-pregame-screen')).toBeInTheDocument();
-      });
-    });
-
```
Comment (I3uckwheat) [inline frontend/src/screenControllers/HostScreenController/HostScreenController.spec.js:118]: Even better

**Review by I3uckwheat — DISMISSED** [review]

**Review by pklepa — DISMISSED** [review]

**Review by zachmmeyer — APPROVED** [review]
Looks great. Tested with a bunch of new players while joining white card and czar selection. No crashing. Only had to update 1 test from the merge.

**Review by I3uckwheat — APPROVED** [review]
